README.md
# NLG Tasks Plugin

![Build status](https://github.com/dataiku/dss-plugin-nlp-nlg-tasks/actions/workflows/auto-make.yml/badge.svg) ![GitHub release (latest by date)](https://img.shields.io/github/v/release/dataiku/dss-plugin-nlp-nlg-tasks?logo=github) ![Support level](https://img.shields.io/badge/support-Unsupported-orange)

This Dataiku DSS plugin provides a recipe to perform any kind of natural language task on a text column using Natural Language Generation (NLG) models, such as [OpenAI's GPT-3 model](https://openai.com/blog/openai-api/).

## Release notes

See the [changelog](CHANGELOG.md) for a history of notable changes to this plugin.

## License

This plugin is distributed under the [Apache License version 2.0](LICENSE).


CHANGELOG.md
# Changelog

## [Version 1.0.0](https://github.com/dataiku/dss-plugin-nlp-nlg-tasks/releases/tag/v1.0.0) - Initial release - 2021-12

- ✨ Release of the plugin with support for OpenAIs GPT APIs


python-lib/parallelizer.py
# -*- coding: utf-8 -*-
"""Apply a function to a pandas DataFrame with parallelization, error logging and progress tracking"""

### TODO (Niklas): Replace with subrepo as soon as parallelizer is merged ###

import logging
import inspect
import math

from collections import namedtuple
from collections import OrderedDict
from concurrent.futures import as_completed
from concurrent.futures import ThreadPoolExecutor
from copy import deepcopy
from enum import Enum
from time import perf_counter
from typing import Any
from typing import AnyStr
from typing import Callable
from typing import Dict
from typing import List
from typing import NamedTuple
from typing import Optional
from typing import Tuple
from typing import Union

from more_itertools import chunked
from more_itertools import flatten
import pandas as pd
from tqdm.auto import tqdm as tqdm_auto

from plugin_io_utils import generate_unique


class ErrorHandling(Enum):
    """Enum class to identify how to handle API errors"""

    LOG = "Log"
    FAIL = "Fail"


class BatchError(ValueError):
    """Custom exception raised if the Batch function fails"""


def _parse_batch_response_default(
    batch: List[Dict], response: Any, output_column_names: NamedTuple
) -> List[Dict]:
    """Adds the response column to the row dictionary at batch[0], while keeping the
    existing dict entries. Should only be used when batch_size=1.
    Args:
        batch: Single input row from the dataframe as a dict in a list of length 1
        response: Response returned by the API, typically a JSON string
        output_column_names: Column names to be added to the row,
            as defined in _get_unique_output_column_names
    Returns:
        batch: Same as input batch with additional columns
            corresponding to the default output columns
    """
    return [
        {
            output_column_names.response: response,
            output_column_names.error_message: "",
            output_column_names.error_type: "",
            output_column_names.error_raw: "",
            **batch[0],
        }
    ]


class DataFrameParallelizer:
    """Apply a function to a pandas DataFrame with parallelization, error logging and progress tracking.
    This class is particularly well-suited for synchronous functions calling an API, either row-by-row or by batch.
    Attributes:
        function: Any function taking a dict as input (row-by-row mode) or a list of dict (batch mode),
            and returning a response with additional information, typically a JSON string.
            In batch mode, the response from the function should be parsable by the `batch_response_parser` attribute.
        error_handling: If ErrorHandling.LOG (default), log the error from the function as a warning,
            and add additional columns to the dataframe with the error message and error type.
            If ErrorHandling.FAIL, the function will fail is there is any error.
            We recommend letting the end user choose as there are contexts which justify one option or the other.
        exceptions_to_catch: Tuple of Exception classes to catch.
            Mandatory if ErrorHandling.LOG (default).
        parallel_workers: Number of concurrent threads to parallelize the function. Default is 4.
            We recommend letting the end user tune this parameter to get better performance.
        batch_support: If True, send batches of row (list of dict) to the `function`
            Else (default) send rows as dict to the function.
            This parameter should be chosen according to the nature of the function to apply.
        batch_size: Number of rows to include in each batch. Default is 10.
            Taken into account if `batch_support` is True.
            We recommend letting the end user tune this parameter if they need to increase performance.
        batch_response_parser: Function used to parse the raw response from the function in batch mode,
            and assign the actual responses and errors back to the original batch of row (list of dict).
            This is often required for batch APIs which return nested objects with a mix of responses and errors.
            This parameter is required if batch_support is True.
        output_column_prefix: Column prefix to add to the output columns of the dataframe,
            containing the `function` responses and errors. Default is "output".
            This should be overriden by the developer: if the function to apply calls an API for text translation,
            a good output_column_prefix would be "api_translation".
        verbose: If True, log raw details on any error encountered along with the error message and error type.
            Else (default) log only the error message and the error type.
            We recommend trying without verbose first. Usually, the error message is enough to diagnose the issue.
    """

    # Default number of worker threads to use in parallel - may be tuned by the end user
    DEFAULT_PARALLEL_WORKERS = 4
    # By default, we assume the function to apply is row-by-row - should be overriden in the batch case
    DEFAULT_BATCH_SUPPORT = False
    # Default number of rows in one batch - may be tuned by the end user
    DEFAULT_BATCH_SIZE = 10
    # Default response parsing function for batch_size=1 - Simply assigns the response to the response column
    DEFAULT_RESPONSE_PARSER = _parse_batch_response_default
    # Default prefix to add to output columns - should be overriden for personalized output
    DEFAULT_OUTPUT_COLUMN_PREFIX = "output"
    # Default dictionary of output column names (key) and their descriptions (value)
    OUTPUT_COLUMN_NAME_DESCRIPTIONS = OrderedDict(
        [
            ("response", "Raw response in JSON format"),
            ("error_message", "Error message"),
            ("error_type", "Error type or code"),
            ("error_raw", "Raw error"),
        ]
    )
    # By default, set verbose to False assuming error message and type are enough information in the logs
    DEFAULT_VERBOSE = False

    def __init__(
        self,
        function: Callable[[Union[Dict, List[Dict]]], Union[Dict, List[Dict]]],
        error_handling: ErrorHandling = ErrorHandling.LOG,
        exceptions_to_catch: Tuple[Exception] = (),
        parallel_workers: int = DEFAULT_PARALLEL_WORKERS,
        batch_support: bool = DEFAULT_BATCH_SUPPORT,
        batch_size: int = DEFAULT_BATCH_SIZE,
        batch_response_parser: Callable[
            [List[Dict], Any, NamedTuple], List[Dict]
        ] = DEFAULT_RESPONSE_PARSER,
        output_column_prefix: AnyStr = DEFAULT_OUTPUT_COLUMN_PREFIX,
        verbose: bool = DEFAULT_VERBOSE,
    ):
        self.function = function
        self.error_handling = error_handling
        self.exceptions_to_catch = exceptions_to_catch
        if error_handling == ErrorHandling.LOG and not exceptions_to_catch:
            raise ValueError("Please set at least one exception in exceptions_to_catch")
        self.parallel_workers = parallel_workers
        self.batch_support = batch_support
        if not batch_support:
            # Overwrite necessary args for row-by-row iteration
            batch_size = 1
            batch_response_parser = _parse_batch_response_default

        self.batch_size = batch_size
        self.batch_response_parser = batch_response_parser
        self.output_column_prefix = output_column_prefix
        self.verbose = verbose
        self._output_column_names = None  # Will be set at runtime by the run method

    def _get_unique_output_column_names(self, existing_names: List[AnyStr]) -> NamedTuple:
        """Return a named tuple with prefixed column names and their descriptions"""
        OutputColumnNameTuple = namedtuple(
            "OutputColumnNameTuple", self.OUTPUT_COLUMN_NAME_DESCRIPTIONS.keys()
        )
        return OutputColumnNameTuple(
            *[
                generate_unique(
                    name=column_name,
                    existing_names=existing_names,
                    prefix=self.output_column_prefix,
                )
                for column_name in OutputColumnNameTuple._fields
            ]
        )

    def _apply_function_and_parse_response(
        self,
        batch: List[Dict] = None,
        **function_kwargs,
    ) -> Union[Dict, List[Dict]]:  # sourcery skip: or-if-exp-identity
        """Wrap a row-by-row or batch function with error logging and response parsing
        It applies `self.function` and:
        - If batch, parse the function response to extract results and errors using `self.batch_response_parser`
        - handles errors from the function with two methods:
            * (default) log the error message as a warning and return the row with error keys
            * fail if there is an error (if `self.error_handling == ErrorHandling.FAIL`)
        """
        output = deepcopy(batch)
        for output_column in self._output_column_names:
            for output_row in output:
                output_row[output_column] = ""
        try:
            if not self.batch_support:
                # In the row-by-row case, there is only one element in the list as batch_size=1
                response = self.function(row=batch[0], **function_kwargs)
            else:
                response = self.function(batch=batch, **function_kwargs)
            output = self.batch_response_parser(
                batch=batch, response=response, output_column_names=self._output_column_names
            )
            errors = [
                row[self._output_column_names.error_message]
                for row in output
                if row[self._output_column_names.error_message]
            ]
            if errors:
                raise BatchError(str(errors))
        except self.exceptions_to_catch + (BatchError,) as error:
            if self.error_handling == ErrorHandling.FAIL:
                raise error
            logging.warning(
                f"Function {self.function.__name__} failed on: {batch} because of error: {error}"
            )
            error_type = str(type(error).__qualname__)
            module = inspect.getmodule(error)
            if module:
                error_type = f"{module.__name__}.{error_type}"
            for output_row in output:
                output_row[self._output_column_names.error_message] = str(error)
                output_row[self._output_column_names.error_type] = error_type
                output_row[self._output_column_names.error_raw] = str(error.args)
        return output

    def _convert_results_to_df(self, df: pd.DataFrame, results: List[Dict]) -> pd.DataFrame:
        """Combine results from the function with the input dataframe"""
        output_schema = {
            **{column_name: str for column_name in self._output_column_names},
            **dict(df.dtypes),
        }
        output_df = (
            pd.DataFrame.from_records(results)
            .reindex(columns=list(df.columns) + list(self._output_column_names))
            .astype(output_schema)
        )
        if not self.verbose:
            output_df.drop(labels=self._output_column_names.error_raw, axis=1, inplace=True)
        if self.error_handling == ErrorHandling.FAIL:
            error_columns = [
                self._output_column_names.error_message,
                self._output_column_names.error_type,
                self._output_column_names.error_raw,
            ]
            output_df.drop(labels=error_columns, axis=1, inplace=True, errors="ignore")
        return output_df

    def run(
        self,
        df: pd.DataFrame,
        **function_kwargs,
    ) -> pd.DataFrame:
        """Apply a function to a pandas.DataFrame with parallelization, error logging and progress tracking.
        The DataFrame is iterated on and fed to the function as dictionaries, row-by-row or by batches of rows.
        This process is accelerated by the use of concurrent threads and is tracked with a progress bar.
        Errors are catched if they match the `self.exceptions_to_catch` attribute and automatically logged.
        Once the whole DataFrame has been iterated on, results and errors are added as additional columns.
        Args:
            df: Input dataframe on which the function will be applied
            **function_kwargs: Arbitrary keyword arguments passed to the `function`
        Returns:
            Input dataframe with additional columns:
            - response from the `function`
            - error message if any
            - error type if any
        """
        # First, we create a generator expression to yield each row of the input dataframe.
        # Each row will be represented as a dictionary like {"column_name_1": "foo", "column_name_2": 42}
        df_row_generator = (index_series_pair[1].to_dict() for index_series_pair in df.iterrows())
        # We then chunk the generator into lists of rows of len batch_size
        df_row_batch_generator = chunked(df_row_generator, self.batch_size)
        df_num_rows = len(df.index)
        len_generator = math.ceil(df_num_rows / self.batch_size)
        logging.info(
            f"Applying function {self.function.__name__} in parallel to {df_num_rows} row(s)"
            + f" using batch size of {self.batch_size}..."
        )
        start = perf_counter()
        self._output_column_names = self._get_unique_output_column_names(existing_names=df.columns)
        pool_kwargs = function_kwargs.copy()
        for kwarg in ["function", "row", "batch"]:  # Reserved pool keyword arguments
            pool_kwargs.pop(kwarg, None)
        results = []
        with ThreadPoolExecutor(max_workers=self.parallel_workers) as pool:
            futures = [
                pool.submit(self._apply_function_and_parse_response, batch=batch, **pool_kwargs)
                for batch in df_row_batch_generator
            ]
            for future in tqdm_auto(
                as_completed(futures), total=len_generator, miniters=1, mininterval=1.0
            ):
                results.append(future.result())
        results = flatten(results)
        output_df = self._convert_results_to_df(df, results)
        num_error = sum(output_df[self._output_column_names.response] == "")
        num_success = len(df.index) - num_error
        logging.info(
            (
                f"Applied function in parallel: {num_success} row(s) succeeded, {num_error} failed "
                f"in {(perf_counter() - start):.2f} seconds."
            )
        )
        return output_df


python-lib/gpt_api_client.py
# -*- coding: utf-8 -*-
"""Module with client calling the OpenAI GPT completion endpoint"""

from typing import List
from typing import Tuple

import openai
import requests

# ==============================================================================
# CONSTANT DEFINITION
# ==============================================================================

API_EXCEPTIONS = (requests.HTTPError,)

# ==============================================================================
# CLASS AND FUNCTION DEFINITION
# ==============================================================================


class GPTClient:
    def __init__(self, engine, api_key) -> None:
        self.engine = engine
        openai.api_key = api_key

    def format_prompt(
        self,
        task: str = "",
        text: str = "",
        input_desc: str = "",
        output_desc: str = "",
        examples: List[Tuple[str, str]] = [("", "")],
    ) -> str:
        """
        Returns prompt of form:

        Correct grammar mistakes.

        Original: Where do you went?
        Standard American English: Where did you go?
        Original: Where is you?
        Standard American English:

        Args:
            task: The task for GPT, e.g. Correct grammar mistakes.
            text: The current row based on which to generate
            input_desc: Description of input column
            output_desc: Description of output column
            example_in: Example of input text
            example_out: Example of desired output text
        Returns:
            prompt: Formatted prompt
        """

        prompt = ""

        ### Task ###
        if task:
            prompt += f"{task}\n\n"

        ### Examples ###
        for ex_inp, ex_out in examples:
            if ex_inp:
                if input_desc:
                    prompt += f"{input_desc}: "
                prompt += f"{ex_inp}\n"

            # One could also provide output examples without descriptions, e.g.
            # elephant
            # giraffe
            # cat
            if ex_out:
                if output_desc:
                    prompt += f"{output_desc}: "
                prompt += f"{ex_out}\n"

        ### Final prompt ###
        if text:
            if input_desc:
                prompt += f"{input_desc}: "
            prompt += f"{text}\n"

        if output_desc:
            # Do not end with a space, as it worsens generation
            prompt += f"{output_desc}:"

        return prompt

    def generate(
        self,
        task: str = "",
        text: str = "",
        input_desc: str = "",
        output_desc: str = "",
        examples: List[Tuple[str, str]] = [("", "")],
        temperature: float = 0.7,
        max_tokens: int = 64,
    ) -> str:
        """
        Constructs a prompt and makes an API call to generate text.
        Default values for temperature and max_tokens are chosen based on the OpenAI playground.
        """
        prompt = self.format_prompt(task, text, input_desc, output_desc, examples)

        response = openai.Completion.create(
            engine=self.engine,
            prompt=prompt,
            stop="\n",
            temperature=temperature,
            max_tokens=max_tokens,
        )

        if "choices" in response:
            return response["choices"][0]
        else:
            # OpenAIs Python client seems to handle all exceptions so this should rarely be called
            user_message = f"Encountered the following error while sending an API request to OpenAI: {response}"
            raise requests.HTTPError(user_message)


python-lib/plugin_io_utils.py
# -*- coding: utf-8 -*-
"""Module with read/write utility functions which are *not* based on the Dataiku API"""

from collections import namedtuple
from collections import OrderedDict
import json
import logging
from enum import Enum
from typing import AnyStr
from typing import Dict
from typing import List
from typing import NamedTuple

import pandas as pd

# ==============================================================================
# CONSTANT DEFINITION
# ==============================================================================

COLUMN_PREFIX = "api"
API_COLUMN_NAMES_DESCRIPTION_DICT = OrderedDict(
    [
        ("response", "Raw response from the API in JSON format"),
        ("error_message", "Error message from the API"),
        ("error_type", "Error type (module and class name)"),
        ("error_raw", "Raw error from the API"),
    ]
)

ApiColumnNameTuple = namedtuple("ApiColumnNameTuple", API_COLUMN_NAMES_DESCRIPTION_DICT.keys())


class ErrorHandlingEnum(Enum):
    LOG = "Log"
    FAIL = "Fail"


# ==============================================================================
# CLASS AND FUNCTION DEFINITION
# ==============================================================================


def generate_unique(name: AnyStr, existing_names: List, prefix: AnyStr = None) -> AnyStr:
    """
    Generate a unique name among existing ones by suffixing a number. Can also add an optional prefix.
    """
    if prefix is not None:
        base_name = prefix + "_" + name
    else:
        base_name = name

    new_name = base_name
    for j in range(1, 1001):
        if new_name not in existing_names:
            return new_name
        new_name = base_name + "_{}".format(j)
    raise Exception("Failed to generated a unique name")


def build_unique_column_names(
    existing_names: List[AnyStr], column_prefix: AnyStr = COLUMN_PREFIX
) -> NamedTuple:
    """
    Helper function to the "api_parallelizer" main function.
    Initializes a named tuple of column names from ApiColumnNameTuple, ensure columns are unique.
    """
    api_column_names = ApiColumnNameTuple(
        *[generate_unique(k, existing_names, column_prefix) for k in ApiColumnNameTuple._fields]
    )
    return api_column_names


def validate_column_input(column_name: AnyStr, column_list: List[AnyStr]) -> None:
    """
    Validate that user input for column parameter is valid.
    """
    if column_name is None or len(column_name) == 0:
        raise ValueError("You must specify a valid column name.")
    if column_name not in column_list:
        raise ValueError("Column '{}' is not present in the input dataset.".format(column_name))


def safe_json_loads(
    str_to_check: AnyStr,
    error_handling: ErrorHandlingEnum = ErrorHandlingEnum.LOG,
    verbose: bool = False,
) -> Dict:
    """
    Wrap json.loads with an additional parameter to handle errors:
    - 'FAIL' to use json.loads, which throws an exception on invalid data
    - 'LOG' to try json.loads and return an empty dict if data is invalid
    """
    if error_handling == ErrorHandlingEnum.FAIL:
        output = json.loads(str_to_check)
    else:
        try:
            output = json.loads(str_to_check)
        except (TypeError, ValueError):
            if verbose:
                logging.warning("Invalid JSON: '" + str(str_to_check) + "'")
            output = {}
    return output


def move_api_columns_to_end(
    df: pd.DataFrame,
    api_column_names: NamedTuple,
    error_handling: ErrorHandlingEnum = ErrorHandlingEnum.LOG,
) -> pd.DataFrame:
    """
    Move non-human-readable API columns to the end of the dataframe
    """
    api_column_names_dict = api_column_names._asdict()
    if error_handling == ErrorHandlingEnum.FAIL:
        api_column_names_dict.pop("error_message", None)
        api_column_names_dict.pop("error_type", None)
    if not any(["error_raw" in k for k in df.keys()]):
        api_column_names_dict.pop("error_raw", None)
    cols = [c for c in df.keys() if c not in api_column_names_dict.values()]
    new_cols = cols + list(api_column_names_dict.values())
    df = df.reindex(columns=new_cols)
    return df


python-lib/gpt_api_formatting.py
# -*- coding: utf-8 -*-
"""Module with classes to format results from the OpenAI GPT completion endpoint"""

import logging
from typing import AnyStr
from typing import Dict

import pandas as pd
from plugin_io_utils import (
    API_COLUMN_NAMES_DESCRIPTION_DICT,
    ErrorHandlingEnum,
    build_unique_column_names,
    generate_unique,
    safe_json_loads,
    move_api_columns_to_end,
)

# ==============================================================================
# CLASS AND FUNCTION DEFINITION
# ==============================================================================


class GenericAPIFormatter:
    """
    Generic Formatter class for API responses:
    - initialize with generic parameters
    - compute generic column descriptions
    - apply format_row to dataframe
    """

    def __init__(
        self,
        input_df: pd.DataFrame,
        column_prefix: AnyStr = "api",
        error_handling: ErrorHandlingEnum = ErrorHandlingEnum.LOG,
    ):
        self.input_df = input_df
        self.column_prefix = column_prefix
        self.error_handling = error_handling
        self.api_column_names = build_unique_column_names(input_df, column_prefix)
        self.column_description_dict = {
            v: API_COLUMN_NAMES_DESCRIPTION_DICT[k]
            for k, v in self.api_column_names._asdict().items()
        }

    def format_row(self, row: Dict) -> Dict:
        return row

    def format_df(self, df: pd.DataFrame) -> pd.DataFrame:
        logging.info("Formatting API results...")
        df = df.apply(func=self.format_row, axis=1)
        df = move_api_columns_to_end(df, self.api_column_names, self.error_handling)
        logging.info("Formatting API results: Done.")
        return df


class GPTAPIFormatter(GenericAPIFormatter):
    """
    Formatter class for GPT API responses.
    Response has to be a valid JSON.
    """

    def __init__(
        self,
        input_df: pd.DataFrame,
        input_column: AnyStr = "",
        output_column: AnyStr = "generation",
        column_prefix: AnyStr = "gpt",
        response_column: AnyStr = "text",
        output_only_mode: bool = False,
        error_handling: ErrorHandlingEnum = ErrorHandlingEnum.LOG,
    ):
        super().__init__(input_df, column_prefix, error_handling)

        if output_only_mode:
            self.generated_text_column_name = output_column
        else:
            self.generated_text_column_name = generate_unique(
                output_column, input_df.columns, prefix=None
            )
        self.output_only_mode = output_only_mode
        self.input_column = input_column
        self.input_df_columns = input_df.columns
        self.response_column = response_column
        self._compute_column_description()

    def _compute_column_description(self):
        if self.output_only_mode:
            self.column_description_dict[self.generated_text_column_name] = "Generated text"
        else:
            self.column_description_dict[
                self.generated_text_column_name
            ] = f'Generation based on "{self.input_column}" column'

    def format_row(self, row: Dict) -> Dict:
        """
        Formats raw row with response into final dataframe row.

        Args:
            row: Dict of a single dataframe row with a column corresponding to the response.
        Returns:
            row: Dict of a single formatted dataframe row
        """
        raw_response = row[self.api_column_names.response]
        response = safe_json_loads(raw_response, self.error_handling)
        row[self.generated_text_column_name] = response.get(self.response_column, "")
        return row


custom-recipes/nlp-gpt-tasks/recipe.py
# -*- coding: utf-8 -*-
import json
from typing import Dict
from typing import List
from typing import Tuple

import pandas as pd
from retry import retry

import dataiku
from dataiku.customrecipe import get_input_names_for_role
from dataiku.customrecipe import get_output_names_for_role
from dataiku.customrecipe import get_recipe_config

from dkulib.dku_io_utils import set_column_descriptions
from dkulib.parallelizer import DataFrameParallelizer
from gpt_api_client import API_EXCEPTIONS
from gpt_api_client import GPTClient
from gpt_api_formatting import GPTAPIFormatter
from plugin_io_utils import ErrorHandlingEnum
from plugin_io_utils import validate_column_input

# ==============================================================================
# SETUP
# ==============================================================================

recipe_config = get_recipe_config()

api_configuration_preset = recipe_config.get("api_configuration_preset")
if api_configuration_preset is None or api_configuration_preset == {}:
    raise ValueError("Please specify an API configuration preset")

output_only_mode = recipe_config.get("output_only_mode", False)
if output_only_mode:
    examples = recipe_config.get("output_examples", "")
    examples = [("", v) for v in examples]
    # Explicity set to empty strings as DSS may cache previous settings
    input_desc = ""
    text_column = ""
else:
    examples = recipe_config.get("examples")
    examples = [(k, v) for k, v in examples.items()]
    input_desc = recipe_config.get("input_desc", "")
    # If none specificed, will trigger <class 'ValueError'>: You must specify a valid column name
    text_column = recipe_config.get("text_column")

task = recipe_config.get("task", "")
output_desc = recipe_config.get("output_desc", "")
temperature = recipe_config.get("temperature", 0.7)
max_tokens = recipe_config.get("max_tokens", 64)

# Create a fitting name for the output column
if output_desc:
    output_column_name = output_desc.lower().replace(" ", "_")
else:
    output_column_name = "generation"

# Params for parallelization
column_prefix = "gpt"
parallel_workers = api_configuration_preset.get("parallel_workers")
error_handling = (
    ErrorHandlingEnum.FAIL if get_recipe_config().get("fail_on_error") else ErrorHandlingEnum.LOG
)

# Create client
client = GPTClient(api_configuration_preset.get("engine"), api_configuration_preset.get("api_key"))
max_attempts = api_configuration_preset.get("max_attempts")
wait_interval = api_configuration_preset.get("wait_interval")


# ==============================================================================
# DEFINITIONS
# ==============================================================================

if output_only_mode:
    input_dataset = None
    # Simulate an empty input dataframe if none is specified
    input_df = pd.DataFrame([""] * recipe_config.get("num_outputs"), columns=[output_column_name])
else:
    input_dataset_names = get_input_names_for_role("input_dataset")
    if not input_dataset_names:
        raise ValueError(
            "Cannot find input dataset. Use Output-only mode to generate without input dataset."
        )
    input_dataset = dataiku.Dataset(input_dataset_names[0])
    validate_column_input(text_column, [col["name"] for col in input_dataset.read_schema()])
    input_df = input_dataset.get_dataframe()

output_dataset = dataiku.Dataset(get_output_names_for_role("output_dataset")[0])


@retry((API_EXCEPTIONS), delay=wait_interval, tries=max_attempts)
def call_gpt_api(
    row: Dict,
    text_column: str = "",
    task: str = "",
    input_desc: str = "",
    output_desc: str = "",
    examples: List[Tuple[str, str]] = [("", "")],
    temperature: float = 0.7,
    max_tokens: int = 64,
) -> str:
    """
    Calls GPT Text Generation API.
    """
    if text_column:
        text = row[text_column]
    else:
        text = ""

    # Recipe UI will show an error when selecting a non-string input column
    if not isinstance(text, str):
        return json.dumps({})
    else:
        response = client.generate(
            task=task,
            text=text,
            input_desc=input_desc,
            output_desc=output_desc,
            examples=examples,
            temperature=temperature,
            max_tokens=max_tokens,
        )
        return response


formatter = GPTAPIFormatter(
    input_df=input_df,
    output_column=output_column_name,
    input_column=text_column,
    column_prefix=column_prefix,
    output_only_mode=output_only_mode,
    error_handling=error_handling,
)

# ==============================================================================
# RUN
# ==============================================================================

df_parallelizer = DataFrameParallelizer(
    function=call_gpt_api,
    error_handling=error_handling,
    exceptions_to_catch=API_EXCEPTIONS,
    parallel_workers=parallel_workers,
    output_column_prefix=column_prefix,
)

df = df_parallelizer.run(
    input_df,
    text_column=text_column,
    task=task,
    input_desc=input_desc,
    output_desc=output_desc,
    examples=examples,
    temperature=temperature,
    max_tokens=max_tokens,
)

output_df = formatter.format_df(df)
output_dataset.write_with_schema(output_df)

set_column_descriptions(
    input_dataset=input_dataset,
    output_dataset=output_dataset,
    column_descriptions=formatter.column_description_dict,
)
